File: sdk/moltworker/src/credentials.ts
/**
 * FHE Credential Store for Cloudflare Workers
 * 
 * Stores FHE-encrypted credentials in Cloudflare KV with threshold authorization.
 */

import type { Env, Credential } from './types.js';

interface CredentialStoreConfig {
  network: 'helium' | 'nitrogen';
  contractAddress: string;
  kv?: KVNamespace;
}

export class FHECredentialStore {
  private config: CredentialStoreConfig;
  private credentials: Map<string, Credential> = new Map();

  constructor(config: CredentialStoreConfig) {
    this.config = config;
  }

  /**
   * Store an FHE-encrypted credential
   */
  async store(key: string, value: string, threshold = 2): Promise<string> {
    const id = `cred_${key}_${Date.now()}`;
    
    const credential: Credential = {
      id,
      encryptedKey: this.fheEncrypt(key),
      encryptedValue: this.fheEncrypt(value),
      threshold,
      owner: this.config.contractAddress,
      createdAt: Date.now(),
    };

    this.credentials.set(id, credential);

    // Persist to KV if available
    if (this.config.kv) {
      await this.config.kv.put(id, JSON.stringify(credential));
    }

    return id;
  }

  /**
   * Retrieve credential with threshold authorization
   */
  async retrieve(id: string, permits = 2): Promise<string | null> {
    const credential = this.credentials.get(id) || await this.loadFromKV(id);
    
    if (!credential) {
      return null;
    }

    if (permits < credential.threshold) {
      throw new Error(`Insufficient permits: need ${credential.threshold}, got ${permits}`);
    }

    // In production, threshold decryption via Fhenix CoFHE
    // Mock: return mock decrypted value
    return `decrypted_${id}`;
  }

  /**
   * List all credentials
   */
  async list(): Promise<Array<{ id: string; threshold: number }>> {
    const list: Array<{ id: string; threshold: number }> = [];
    
    for (const [id, cred] of this.credentials) {
      list.push({ id, threshold: cred.threshold });
    }

    return list;
  }

  /**
   * Delete a credential
   */
  async delete(id: string): Promise<void> {
    this.credentials.delete(id);
    
    if (this.config.kv) {
      await this.config.kv.delete(id);
    }
  }

  /**
   * Load credential from KV
   */
  private async loadFromKV(id: string): Promise<Credential | null> {
    if (!this.config.kv) return null;

    const data = await this.config.kv.get(id);
    if (!data) return null;

    return JSON.parse(data) as Credential;
  }

  /**
   * Mock FHE encryption
   * In production, uses Fhenix CoFHE
   */
  private fheEncrypt(data: string): string {
    const hash = data.split('').reduce((acc, c) => {
      return ((acc << 5) - acc + c.charCodeAt(0)) | 0;
    }, 0);
    return `0x${Math.abs(hash).toString(16).padStart(64, '0')}`;
  }
}

File: sdk/go/claw/credentials.go
// FHE Credential Vault for GoClaw

package claw

import (
	"fmt"
	"sync"
)

// Credential represents an FHE-encrypted credential
type Credential struct {
	ID             string
	EncryptedKey   []byte
	EncryptedValue []byte
	Threshold      uint8
	Owner          [20]byte
	CreatedAt      uint64
}

// CredentialVault stores FHE-encrypted credentials
type CredentialVault struct {
	address   [20]byte
	threshold uint8
	mu        sync.RWMutex
	store     map[string]Credential
}

// NewCredentialVault creates a new FHE credential vault
func NewCredentialVault(address [20]byte, threshold uint8) *CredentialVault {
	return &CredentialVault{
		address:   address,
		threshold: threshold,
		store:     make(map[string]Credential),
	}
}

// Store stores an FHE-encrypted credential
func (v *CredentialVault) Store(key, value string) (string, error) {
	v.mu.Lock()
	defer v.mu.Unlock()

	if key == "" {
		return "", fmt.Errorf("key cannot be empty")
	}

	id := fmt.Sprintf("cred_%s_%d", key, len(v.store))

	cred := Credential{
		ID:             id,
		EncryptedKey:   FHEEncrypt(key),
		EncryptedValue: FHEEncrypt(value),
		Threshold:      v.threshold,
		Owner:          v.address,
		CreatedAt:      uint64(len(v.store)),
	}

	v.store[id] = cred
	return id, nil
}

// Retrieve retrieves a credential with threshold authorization
func (v *CredentialVault) Retrieve(id string, permits uint8) (string, error) {
	v.mu.RLock()
	defer v.mu.RUnlock()

	if permits < v.threshold {
		return "", fmt.Errorf("insufficient permits: need %d, got %d", v.threshold, permits)
	}

	cred, exists := v.store[id]
	if !exists {
		return "", fmt.Errorf("credential not found: %s", id)
	}

	// FHE decryption would happen here via Fhenix CoFHE
	// Mock: return decrypted value
	return FHEDecrypt(cred.EncryptedValue), nil
}

// List returns all credentials
func (v *CredentialVault) List() []CredentialInfo {
	v.mu.RLock()
	defer v.mu.RUnlock()

	infos := make([]CredentialInfo, 0, len(v.store))
	for id, cred := range v.store {
		infos = append(infos, CredentialInfo{
			ID:        id,
			Key:       string(FHEDecrypt(cred.EncryptedKey)),
			Threshold: cred.Threshold,
		})
	}
	return infos
}

// Delete deletes a credential
func (v *CredentialVault) Delete(id string) error {
	v.mu.Lock()
	defer v.mu.Unlock()

	if _, exists := v.store[id]; !exists {
		return fmt.Errorf("credential not found")
	}
	delete(v.store, id)
	return nil
}
